src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,175 @@
+    const capabilities = mergeCapabilities(defaultCapabilties, {
+      // Get legacy nav links
+      navLinks: Object.assign(
+        {},
+        ...this.http.server.getUiNavLinks().map(({ _id }) => ({ [_id]: true }))
+      ),
+    });
+
+    return modifiers.reduce(
+      async (resolvedCapabilties, modifier) => modifier(this.request, await resolvedCapabilties),
+      Promise.resolve(capabilities)
+    );
+  }
+
+  private async getVariables(app: App, basePath: string, injectedOverrides: Record<string, any>) {
+    const providers: Provider[] = (this.legacy.uiExports as any).defaultInjectedVarProviders || [];
+    const replacers: Replacer[] = (this.legacy.uiExports as any).injectedVarsReplacers || [];
+    const defaultInjectedVars = providers.reduce(
+      (defaults, { fn, pluginSpec }) =>
+        mergeVariables(
+          defaults,
+          fn(this.http.server, pluginSpec.readConfigValue(this.http.server.config, []))
+        ),
+      {}
+    );
+    const appInjectedVars = await this.http.server.getInjectedUiAppVars(app.getId());
+    const injectedVars = mergeVariables(defaultInjectedVars, appInjectedVars, injectedOverrides);
+    const vars = await replacers.reduce(
+      async (variables, replacer) => await replacer(variables, this.request, this.http.server),
+      Promise.resolve(injectedVars)
+    );
+
+    return {
+      strictCsp: this.http.csp.strict,
+      uiPublicUrl: `${basePath}/ui`,
+      bootstrapScriptUrl: `${basePath}/bundles/app/${app.getId()}/bootstrap.js`,
+      i18n: i18n.translate,
+      locale: i18n.getLocale(),
+      darkMode: this.uiSettings?.user?.['theme:darkMode']?.userValue ?? false,
+      injectedMetadata: {
+        version: this.http.server.version,
+        buildNumber: this.env.packageInfo.buildNum,
+        branch: this.env.packageInfo.branch,
+        basePath,
+        env: this.env,
+        legacyMode: app.getId() !== 'core',
+        i18n: {
+          translationsUrl: `${basePath}/translations/${i18n.getLocale()}.json`,
+        },
+        csp: {
+          warnLegacyBrowsers: this.http.csp.warnLegacyBrowsers,
+        },
+        vars,
+        uiPlugins: [...this.plugins.uiPlugins.public].map(([id, plugin]) => ({ id, plugin })),
+        legacyMetadata: {
+          app,
+          bundleId: `app:${app.getId()}`,
+          nav: this.http.server.getUiNavLinks(),
+          env: this.env,
+          version: this.env.packageInfo.version,
+          branch: this.env.packageInfo.branch,
+          buildNum: this.env.packageInfo.buildNum,
+          buildSha: this.env.packageInfo.buildSha,
+          serverName: this.http.server.name,
+          devMode: this.env.mode.dev,
+          basePath,
+          uiSettings: this.uiSettings,
+        },
+        capabilities: await this.getCapabilities(),
+      },
+    };
+  }
+
+  public async render(appId: string, injectedVarsOverrides: Record<string, any> = {}) {
```

Comment:
Are we providing `renderAppWithDefaultConfig` counterpart in the NP? 
https://github.com/elastic/kibana/blob/master/src/legacy/ui/ui_render/ui_render_mixin.js#L302
LP 'hidden' apps utilize it to render themselves.

Reply:
It wasn't in the original draft of this PR, but `renderAppWithDefaultConfig` is just `renderApp` with the ability to set variable overrides. The method currently allows passing this in, which solves this.